Let's set up an example scenario as perspective for our regression exercises using the Zillow dataset.

As a Codeup data science graduate, you want to show off your skills to the Zillow data science team in hopes of getting an interview for a position you saw pop up on LinkedIn. You thought it might look impressive to build an end-to-end project in which you use some of their Kaggle data to predict property values using some of their available features; who knows, you might even do some feature engineering to blow them away. Your goal is to predict the values of single unit properties using the observations from 2017.

In these exercises, you will complete the first step toward the above goal: acquire and prepare the necessary Zillow data from the zillow database in the Codeup database server.

In [3]:
# import necessary libraries
import pandas as pd
import numpy as np
import os
from tqdm import tqdm

# import personal env file
import env

In [4]:
# Use personally developed function to create a wrangle.py file
# WARNING: FUNCTION WILL OVERWRITE ANY 'wrangle.py' FILE THAT EXISTS
env.deploy_wrangle()

In [5]:
# import newly built wrangle file
import wrangle as w

In [6]:
# create write_contents
# write_contents will hold functions I'm going to add to wrangle.py
write_contents = []

1. [x] Acquire `bedroomcnt`, `bathroomcnt`, `calculatedfinishedsquarefeet`, `taxvaluedollarcnt`, `yearbuilt`, `taxamount`, and `fips` from the `zillow` database for all 'Single Family Residential' properties.

- [x] Use MySQL Workbench to isolate the necessary dataset

In [9]:
# build query
query = """
select bedroomcnt, bathroomcnt, calculatedfinishedsquarefeet, taxvaluedollarcnt, yearbuilt, taxamount, fips
from propertylandusetype
	join properties_2017
		using (propertylandusetypeid)
WHERE propertylandusedesc = ("Single Family Residential")
"""

In [10]:
# use env file to get database url
url = env.get_db_url('zillow')

In [11]:
# use pandas SQL query command to get the DataFrame
df = pd.read_sql(query,url)

In [12]:
df.head()

bedroomcnt  bathroomcnt  calculatedfinishedsquarefeet  taxvaluedollarcnt  \
0         0.0          0.0                           NaN            27516.0   
1         0.0          0.0                           NaN               10.0   
2         0.0          0.0                           NaN               10.0   
3         0.0          0.0                           NaN             2108.0   
4         4.0          2.0                        3633.0           296425.0   

   yearbuilt  taxamount    fips  
0        NaN        NaN  6037.0  
1        NaN        NaN  6037.0  
2        NaN        NaN  6037.0  
3        NaN     174.21  6037.0  
4     2005.0    6941.39  6037.0

In [13]:
w.check_file_exists('zillow.csv',query,url)

Reading from file...


bedroomcnt  bathroomcnt  calculatedfinishedsquarefeet  \
0               0.0          0.0                           NaN   
1               0.0          0.0                           NaN   
2               0.0          0.0                           NaN   
3               0.0          0.0                           NaN   
4               4.0          2.0                        3633.0   
...             ...          ...                           ...   
2152858         4.0          3.0                        2262.0   
2152859         4.0          4.5                        3127.0   
2152860         0.0          0.0                           NaN   
2152861         3.0          2.5                        1974.0   
2152862         4.0          4.0                        2110.0   

         taxvaluedollarcnt  yearbuilt  taxamount    fips  
0                  27516.0        NaN        NaN  6037.0  
1                     10.0        NaN        NaN  6037.0  
2                     10.0        NaN        NaN  6037.0  
3                   2108.0        NaN     174.21  6037.0  
4                 296425.0     2005.0    6941.39  6037.0  
...                    ...        ...        ...     ...  
2152858           960756.0     2015.0   13494.52  6059.0  
2152859           536061.0     2014.0    6244.16  6059.0  
2152860           208057.0        NaN    5783.88  6059.0  
2152861           424353.0     2015.0    5302.70  6059.0  
2152862           554009.0     2014.0    6761.20  6037.0  

[2152863 rows x 7 columns]

In [14]:
# create acquire function to prepare this nicely
def acquire_zillow():
    """
    Function to pull in zillow data. Returns a DataFrame. Also builds local csv file of the dataset.
    
    Parameters:
    -----------
    none
    
    """
    # build query
    query = """
select 
    bedroomcnt,
    bathroomcnt,
    calculatedfinishedsquarefeet,
    taxvaluedollarcnt,
    yearbuilt,
    taxamount,
    fips
from properties_2017
left join propertylandusetype
    using(propertylandusetypeid)
where propertylandusedesc = 'Single Family Residential'
    """
    
    # use env file to get database url
    url = env.get_db_url('zillow')
    filename = 'zillow.csv'
    
    df = w.check_file_exists(filename,query,url)
    
    return df

In [15]:
# erase the df and start over to check acquire_zillow function
df = []
df

[]

In [16]:
df = acquire_zillow()
w.df_info(df)

Reading from file...


nunique   dtypes  isnull     769580
bedroomcnt                         19  float64      11       4.00
bathroomcnt                        38  float64      11       3.00
calculatedfinishedsquarefeet    10580  float64    8484    1556.00
taxvaluedollarcnt              592269  float64     493  434000.00
yearbuilt                         153  float64    9337    1958.00
taxamount                      918838  float64    4442    5451.32
fips                                3  float64       0    6037.00

In [17]:
# append the newly built function to write_contents
write_contents.append('''
# create acquire function to prepare this nicely
def acquire_zillow():
    """
    Function to pull in zillow data. Returns a DataFrame. Also builds local csv file of the dataset.
    
    Parameters:
    -----------
    none
    
    """
    # build query
    query = """
select 
    bedroomcnt,
    bathroomcnt,
    calculatedfinishedsquarefeet,
    taxvaluedollarcnt,
    yearbuilt,
    taxamount,
    fips
from properties_2017
left join propertylandusetype
    using(propertylandusetypeid)
where propertylandusedesc = 'Single Family Residential'

    """
    
    # use env file to get database url
    url = env.get_db_url('zillow')
    filename = 'zillow.csv'
    
    df = w.check_file_exists(filename,query,url)
    
    return df
''')

2. [x] Using your acquired Zillow data, walk through the summarization and cleaning steps in your `wrangle.ipynb` file like we did above. You may handle the missing values however you feel is appropriate and meaningful; remember to document your process and decisions using markdown and code commenting where helpful.

In [19]:
# check out the df using a personal function
w.df_info(df,samples=3)

nunique   dtypes  isnull    422358   1465199  \
bedroomcnt                         19  float64      11      3.00      4.00   
bathroomcnt                        38  float64      11      1.00      2.50   
calculatedfinishedsquarefeet    10580  float64    8484   1064.00   2178.00   
taxvaluedollarcnt              592269  float64     493  57347.00  96113.00   
yearbuilt                         153  float64    9337   1954.00   1964.00   
taxamount                      918838  float64    4442   1458.97   1037.78   
fips                                3  float64       0   6037.00   6111.00   

                                 100471  
bedroomcnt                         4.00  
bathroomcnt                        3.00  
calculatedfinishedsquarefeet    2549.00  
taxvaluedollarcnt             279955.00  
yearbuilt                       1926.00  
taxamount                       3578.95  
fips                            6037.00

In [20]:
# Use isnull().mean() to get percentage ratios
round(df.isnull().mean() * 100,3).astype(str)+'%'

bedroomcnt                      0.001%
bathroomcnt                     0.001%
calculatedfinishedsquarefeet    0.394%
taxvaluedollarcnt               0.023%
yearbuilt                       0.434%
taxamount                       0.206%
fips                              0.0%
dtype: object

Considering the ridiculously small proportion of nulls, I'm going to define a rule that if the percent is below 0.01% (or 0.0001, if I'm remembering that properly) then drop the nulls of that column

In [22]:
# get the columns where their proportions are small
auto_drop = [col for col in df.columns if df[col].isna().mean() < 0.001]
auto_drop

['bedroomcnt', 'bathroomcnt', 'taxvaluedollarcnt', 'fips']

In [23]:
# drop the nulls of these and re-examine the dataset
df2 = df.dropna(subset=auto_drop)
w.df_info(df2,include=True)

nunique   dtypes  isnull    1177083      count  \
bedroomcnt                         19  float64       0       4.00  2152359.0   
bathroomcnt                        38  float64       0       2.00  2152359.0   
calculatedfinishedsquarefeet    10578  float64    8099    2290.00  2144260.0   
taxvaluedollarcnt              592265  float64       0  155371.00  2152359.0   
yearbuilt                         153  float64    8958    1905.00  2143401.0   
taxamount                      918821  float64    3971    2092.76  2148388.0   
fips                                3  float64       0    6037.00  2152359.0   

                                       mean            std      min  \
bedroomcnt                         3.287807       0.953796     0.00   
bathroomcnt                        2.231119       0.998916     0.00   
calculatedfinishedsquarefeet    1862.870784    1222.098248     1.00   
taxvaluedollarcnt             461896.824010  699677.417877     1.00   
yearbuilt                       1960.950824      22.161389  1801.00   
taxamount                       5634.944023    8178.944104     1.85   
fips                            6048.374508      20.428394  6037.00   

                                    25%         50%        75%          max  
bedroomcnt                         3.00       3.000       4.00        25.00  
bathroomcnt                        2.00       2.000       3.00        32.00  
calculatedfinishedsquarefeet    1257.00    1623.000    2208.00    952576.00  
taxvaluedollarcnt             188171.00  327671.000  534527.00  98428909.00  
yearbuilt                       1949.00    1958.000    1976.00      2016.00  
taxamount                       2535.06    4108.965    6414.36   1337755.86  
fips                            6037.00    6037.000    6059.00      6111.00

From the comparison, the nuniques went from:
- calculatedfinishedsquarefeet: 10580 -> 10578 (2 val diff)
- taxvaluedollarcnt: 592269 -> 592265 (4 val diff)
- taxamount: 918838 -> 918821 (17 val diff)

Also the nulls went from:
- calculatedfinishedsquarefeet: 8484 -> 8099 (385 diff)
- yearbuilt: 9337 -> 8958 (379 diff)
- taxamount: 4442 -> 3971 (471 diff)

Next let's see about handling that yearbuilt column so we can convert it into an int later.

In [26]:
# maybe we can fill with the average?
df2.yearbuilt.mean()

1960.9508239475488

Doesn't seem too solid. Plus there is a lot of data. We should be safe just dropping them

In [28]:
df = df.dropna()

In [29]:
w.df_info(df,include=True)

nunique   dtypes  isnull   2140364      count  \
bedroomcnt                         19  float64       0      3.00  2140235.0   
bathroomcnt                        38  float64       0      2.00  2140235.0   
calculatedfinishedsquarefeet    10557  float64       0   1796.00  2140235.0   
taxvaluedollarcnt              588178  float64       0  67015.00  2140235.0   
yearbuilt                         153  float64       0   1964.00  2140235.0   
taxamount                      915868  float64       0   1280.26  2140235.0   
fips                                3  float64       0   6059.00  2140235.0   

                                       mean            std      min  \
bedroomcnt                         3.301419       0.932600     0.00   
bathroomcnt                        2.240352       0.990549     0.00   
calculatedfinishedsquarefeet    1863.193970    1221.754161     1.00   
taxvaluedollarcnt             460641.625164  677157.635675    22.00   
yearbuilt                       1960.967545      22.150563  1801.00   
taxamount                       5616.711322    7814.562798     6.34   
fips                            6048.309556      20.344910  6037.00   

                                    25%        50%        75%          max  
bedroomcnt                         3.00       3.00       4.00        25.00  
bathroomcnt                        2.00       2.00       3.00        32.00  
calculatedfinishedsquarefeet    1258.00    1623.00    2208.00    952576.00  
taxvaluedollarcnt             189166.00  328296.00  534606.00  90188462.00  
yearbuilt                       1949.00    1958.00    1976.00      2016.00  
taxamount                       2540.85    4111.47    6411.93   1078101.87  
fips                            6037.00    6037.00    6059.00      6111.00

In [30]:
# we'll adjust everything to integer later
# For now let's see if the dollar amounts would be okay to convert to int as well
df.taxamount

4           6941.39
6          10244.94
7           7924.68
11            91.60
14           255.17
             ...   
2152856    13877.56
2152858    13494.52
2152859     6244.16
2152861     5302.70
2152862     6761.20
Name: taxamount, Length: 2140235, dtype: float64

In [31]:
# do a complicated thing to verify that we're safe to convert to integer for this column
# converts the series into string, splits each item into a list,
# gets the second item, and then gets all the value counts
df.taxvaluedollarcnt.astype(str).str.split('.').str[1].value_counts()

taxvaluedollarcnt
0    2140235
Name: count, dtype: int64

In [32]:
# automate the above process for the rest of the columns
for col in tqdm(df.columns):
    print(df[col].astype(str).str.split('.').str[1].value_counts())
    print()

 14%|█▍        | 1/7 [00:01<00:08,  1.42s/it]

bedroomcnt
0    2140235
Name: count, dtype: int64



 29%|██▊       | 2/7 [00:02<00:07,  1.46s/it]

bathroomcnt
0     1910201
5      230031
75          3
Name: count, dtype: int64



 43%|████▎     | 3/7 [00:04<00:06,  1.59s/it]

calculatedfinishedsquarefeet
0    2140235
Name: count, dtype: int64



 57%|█████▋    | 4/7 [00:06<00:04,  1.65s/it]

taxvaluedollarcnt
0    2140235
Name: count, dtype: int64



 71%|███████▏  | 5/7 [00:08<00:03,  1.72s/it]

yearbuilt
0    2140235
Name: count, dtype: int64



 86%|████████▌ | 6/7 [00:10<00:01,  1.86s/it]

taxamount
24    28892
4     28877
56    28804
7     28795
26    28795
1     28794
8     28745
34    28738
74    28724
14    28703
02    28702
62    28681
94    28650
12    28645
36    28631
08    28630
98    28626
52    28624
92    28617
48    28616
0     28615
82    28604
76    28604
96    28556
44    28540
84    28540
9     28539
72    28531
46    28515
16    28498
28    28494
58    28475
18    28464
04    28444
64    28440
22    28433
2     28428
88    28415
32    28403
78    28389
66    28377
68    28372
86    28355
3     28328
54    28312
38    28306
6     28297
06    28292
5     28258
42    28240
47    14535
51    14491
45    14485
57    14483
83    14468
11    14410
33    14371
49    14348
39    14344
69    14341
35    14326
03    14325
43    14317
63    14310
15    14293
19    14293
27    14283
93    14278
71    14276
29    14266
91    14266
07    14266
41    14264
37    14261
23    14248
77    14238
89    14234
67    14232
09    14228
81    14226
97    14225
17    14223
05    

100%|██████████| 7/7 [00:12<00:00,  1.73s/it]

fips
0    2140235
Name: count, dtype: int64



From the for loop, apparently taxamount and bathroomcnt have decimals, so they're not safe to convert to int

In [34]:
# get a list of the cols
cols = list(df.columns)
cols.remove('taxamount')
cols.remove('bathroomcnt')
cols

['bedroomcnt',
 'calculatedfinishedsquarefeet',
 'taxvaluedollarcnt',
 'yearbuilt',
 'fips']

In [35]:
# convert columns of this list to integer
for col in cols:
    df[col] = df[col].astype(int)

In [36]:
w.df_info(df)

nunique   dtypes  isnull   1185699
bedroomcnt                         19    int64       0       2.0
bathroomcnt                        38  float64       0       1.0
calculatedfinishedsquarefeet    10557    int64       0     925.0
taxvaluedollarcnt              588178    int64       0  368000.0
yearbuilt                         153    int64       0    1942.0
taxamount                      915868  float64       0    4502.6
fips                                3    int64       0    6037.0

In [37]:
def prepare_zillow(df):
    """
    Cleans the zillow DataFrame
    
    Parameters:
    -----------
    df: DataFrame
        - The zillow DataFrame acquired from Codeup's MySQL database (or from local csv)
    
    """
    # re-assign as a copy to fix warnings
    df = df.copy()
    
    # drop the nulls
    df = df.dropna()
    
    # get a list of the cols and drop the two with decimals
    cols = list(df.columns)
    cols.remove('taxamount')
    cols.remove('bathroomcnt')
    
    # convert columns of this list to integer
    for col in cols:
        df[col] = df[col].astype(int)
    
    return df

In [38]:
# test out 2 functions
df = acquire_zillow()

Reading from file...


In [39]:
df = prepare_zillow(df)

In [40]:
w.df_info(df)

nunique   dtypes  isnull    1471212
bedroomcnt                         19    int64       0       4.00
bathroomcnt                        38  float64       0       2.00
calculatedfinishedsquarefeet    10557    int64       0    1559.00
taxvaluedollarcnt              588178    int64       0  430779.00
yearbuilt                         153    int64       0    1962.00
taxamount                      915868  float64       0    5436.13
fips                                3    int64       0    6037.00

In [41]:
write_contents.append('''
def prepare_zillow(df):
    """
    Cleans the zillow DataFrame
    
    Parameters:
    -----------
    df: DataFrame
        - The zillow DataFrame acquired from Codeup's MySQL database (or from local csv)
    
    """
    # re-assign as a copy to fix warnings
    df = df.copy()
    
    # drop the nulls
    df = df.dropna()
    
    # get a list of the cols and drop the two with decimals
    cols = list(df.columns)
    cols.remove('taxamount')
    cols.remove('bathroomcnt')
    cols.remove('propertylandusedesc')
    
    # convert columns of this list to integer
    for col in cols:
        df[col] = df[col].astype(int)
    
    return df
''')

3. [x] Write a function to split your data into train, validate, and test.

In [43]:
from sklearn.model_selection import train_test_split

In [44]:
# importing function previously made
def split_df(df,strat_var,seed=123):
    """
    Returns three dataframes split from one for use in model training, validation, and testing. Takes two arguments:
        df: any dataframe to be split
        strat_var: the value to stratify on. This value should be a categorical variable.
    
    Function performs two splits, first to primarily make the training set, and the second to make the validate and test sets.
    """
    # Run first split
    train, validate_test = train_test_split(df,
                 train_size=0.60,
                random_state=seed,
                 stratify=df[strat_var]
                )
    
    # Run second split
    validate, test = train_test_split(validate_test,
                test_size=0.50,
                 random_state=seed,
                 stratify=validate_test[strat_var]
                )
    
    return train, validate, test

In [45]:
# this yields an error
# train,validate,test = split_df(df,'taxvaluedollarcnt')

In [46]:
train, validate_test = train_test_split(
    df,
    train_size = 0.6,
    random_state = 123
)

In [47]:
train.shape

(1284141, 7)

In [48]:
validate_test.shape

(856094, 7)

In [49]:
validate, test = train_test_split(
    df,
    train_size = 0.5,
    random_state = 123
)

In [50]:
def split_continuous(df,seed=123):
    """
    Returns three dataframes split from one for use in model training, validation, and testing. 
    
    Function performs two splits, first to primarily make the training set, and the second to make the validate and test sets.
    
    Parameters:
    -----------
    df: DataFrame
        - the prepared dataset to be split
    seed: int, defaut=123
        - optional, a seed value to maintain consistency
    """
    # run first split
    train, validate_test = train_test_split(
        df,
        train_size = 0.6,
        random_state = 123
    )
    
    # run second split
    validate, test = train_test_split(
        df,
        train_size = 0.5,
        random_state = 123
    )
    
    return train,validate,test

In [51]:
train,validate,test = split_continuous(df)

In [52]:
train.head()

bedroomcnt  bathroomcnt  calculatedfinishedsquarefeet  \
462820           5          4.0                          3011   
3880             3          2.0                          1682   
992263           4          2.0                          1433   
311722           3          2.5                          2001   
787985           4          2.0                          2277   

        taxvaluedollarcnt  yearbuilt  taxamount  fips  
462820             710674       2005    7561.14  6111  
3880               603333       1961    7086.19  6037  
992263              82804       1954    1134.83  6037  
311722             355444       1999    4174.92  6111  
787985             571000       1962    7113.02  6037

In [53]:
write_contents.append('''
def split_continuous(df,seed=123):
    """
    Returns three dataframes split from one for use in model training, validation, and testing. 
    
    Function performs two splits, first to primarily make the training set, and the second to make the validate and test sets.
    
    Parameters:
    -----------
    df: DataFrame
        - the prepared dataset to be split
    seed: int, defaut=123
        - optional, a seed value to maintain consistency
    """
    # run first split
    train, validate_test = train_test_split(
        df,
        train_size = 0.6,
        random_state = 123
    )
    
    # run second split
    validate, test = train_test_split(
        df,
        train_size = 0.5,
        random_state = 123
    )
    
    return train,validate,test
''')

4. [x] Store all of the necessary functions to automate your process from acquiring the data to returning a cleaned dataframe with no missing values in your `wrangle.py` file. Name your final function `wrangle_zillow`.

In [55]:
def wrangle_zillow():
    """
    Acquires, prepares, and splits the zillow dataset.
    
    Parameters:
    -----------
    none
    
    
    """
    
    # acquire the data
    zillow = acquire_zillow()
    
    # prepare the data
    zillow = prepare_zillow(zillow)
    
    # split the data
    train,validate,test = split_continuous(zillow)
    
    return train,validate,test

In [56]:
# wipe df clean
df = []
df

[]

In [57]:
train = []
train

[]

In [58]:
tvt_set = train,validate,test = wrangle_zillow()

Reading from file...


In [59]:
w.df_info(train)

nunique   dtypes  isnull     766875
bedroomcnt                         17    int64       0       3.00
bathroomcnt                        37  float64       0       2.00
calculatedfinishedsquarefeet     9540    int64       0    1350.00
taxvaluedollarcnt              446402    int64       0  103146.00
yearbuilt                         149    int64       0    1989.00
taxamount                      707961  float64       0    2089.89
fips                                3    int64       0    6037.00

In [60]:
write_contents.append('''
def wrangle_zillow():
    """
    Acquires, prepares, and splits the zillow dataset.
    
    Parameters:
    -----------
    none
    
    
    """
    
    # acquire the data
    zillow = acquire_zillow()
    
    # prepare the data
    zillow = prepare_zillow(zillow)
    
    # split the data
    train,validate,test = split_continuous(zillow)
    
    return train,validate,test
''')

In [61]:
# save all the functions made to wrangle.py
with open('wrangle.py','a') as file:
    for content in write_contents:
        file.write(content + '\n')